fastssl/models/emp.py

```diff
@@ -0,0 +1,133 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from fastssl.models.ssl import SSL
+from fastssl.models.backbone import BackBone
+from fastssl.data.cifar_transforms import CifarTransform
+from fastssl.data.imagenet_dataloaders import TransformGPU
+
+ssl_transform = CifarTransform()
+
+
+def get_collapse_avoiding_loss(z, criterion, num_patches):
+    z_list = z.chunk(num_patches, dim=0)
+    loss = 0
+    for i in range(num_patches):
+        loss += criterion(z_list[i])
+    loss = loss / num_patches
+    return loss
+
+
+class ClusterCosineSimilarity(nn.Module):
+    def __init__(
+        self,
+    ):
+        super().__init__()
+        pass
+
+    def forward(self, z_proj_clusters, num_patches=2):
+        """
+        Compute cluster centers and similarity loss.
+        Args:
+            z_proj_clusters: list of projected features from different views (NP x B x D)
+            num_patches: (int) number of patches (NP)
+        """
+        z_proj_centers = torch.stack(z_proj_clusters, dim=0).mean(dim=0, keepdim=True)
+        z_sim = F.cosine_similarity(z_proj_centers, z_proj_clusters, dim=-1).mean()
+        # we want to maximize similarity, therefore negative sign
+        return -z_sim
+
+
+class TotalCodingRate(nn.Module):
+    # credits: https://github.com/tsb0601/EMP-SSL/blob/main/main.py
+    def __init__(self, eps=0.2):
+        super(TotalCodingRate, self).__init__()
+        self.eps = eps
+
+    def compute_discrimn_loss(self, W):
+        """Discriminative Loss."""
+        p, m = W.shape  # [d, B]
+        I = torch.eye(p, device=W.device)
+        scalar = p / (m * self.eps)
+        logdet = torch.logdet(I + scalar * W.matmul(W.T))
+        return logdet / 2.0
+
+    def forward(self, X):
+        return -self.compute_discrimn_loss(X.T)
+
+
+def EMPLoss(model, inp, _lambda=None, patch_sim_scale=200, num_patches=2):
+    """
+    Implement Extreme Multi-Patch training.
+    """
+    # build losses
+    simloss = ClusterCosineSimilarity()
+    tcrloss = TotalCodingRate()
+
+    inp_multiview = torch.cat(inp, dim=0)
```

Comment:
Interesting that the effective batchsize for forward propagation now becomes num_patches x bsz?